# 2B-remove-common-variants.sh

SAMPLENAME=PC3 ;
mkdir ./1kg_germline_output ;
THOUSANDGENOMESVCF=/media/sciseim/mrwing/temp/SnpEff/VCF-testing/ensemblFTP/1000GENOMES-phase_3.vcf


# All variants present in the 1000 Genomes data and annotated in Ensembl hg38 build 82 were excluded to obtain only somatic mutation data and exclude common germline variants.
# 
# data file available at 
# bundled variant data file available at \href{ftp.ensembl.org/pub/release-82/variation/vcf/homo_sapiens/1000GENOMES-phase_3.vcf.gz

# ############################################################### 
# split 1000 genomes file (12.5GB) by chromosome
# Split VCF by Chromosome
mkdir ./1kG_split_by_chr && for chr in M 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y; do vcftools --chr $chr --vcf 1000GENOMES-phase_3.vcf --out ./1kG_split_by_chr/$chr.1000GENOMES-phase_3.vcf --recode --keep-INFO-all ;done 
# ... e.g. get M.1000GENOMES-phase_3.vcf.recode.vcf
THOUSANDGENOMESVCFFOLDER=/media/sciseim/mrwing/temp/SnpEff/VCF-testing/ensemblFTP/1kG_split_by_chr
# ############################################################### 


# ############################################################### 
# SPLIT SAMPLE VCF BY CHROMOSOME
# GO TO THE DIRECTORY OF THE VCF GENERATED BY MAPPING YOUR GENES TO THE REFERENCE GENOME
# just split the VCF by chromosome
# Split VCF by Chromosome
# here hg38, so naming is '1', not 'chr1' etc
mkdir ./split_by_chr && for chr in M 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 X Y; do vcftools --chr $chr --vcf PC3-reads.vcf --out ./split_by_chr/$SAMPLENAME-reads.$chr --recode --keep-INFO-all && cd ./split_by_chr && gzip $SAMPLENAME-reads.$chr.recode.vcf && cd .. ;done 
# now have e.g. 'reads.Y.recode.vcf' in the folder 'split_by_chr'
# ############################################################### 

# ############################################################### 
# KEEP VARIANTS LIKELY UNIQUE TO YOUR SAMPLE
#
# i.e. keep single nucleotide variants (SNVs) and short indel variants likely to NOT be germline variants
cd ./split_by_chr ;

for chr in 1 10 11 12 13 14 15 16 17 18 19 2 20 21 22 3 4 5 6 7 8 9 M X Y; do bedtools intersect -header -v -wa -wb -a $SAMPLENAME-reads.$chr.recode.vcf -b $THOUSANDGENOMESVCFFOLDER/$chr.1000GENOMES-phase_3.vcf.recode.vcf -sorted >./1kg-filtered/$chr.1k-filtered.vcf;done 
# ############################################################### 

